### ISSUE TITLE
Add CSS Framework based Themes at Installation

### ISSUE THREAD

**Issue by modir** [contao/contao#6678]
### Description

I have seen here several issues about the default CSS Grid Framework etc. e.g #3235 or #832. I would like to propose a different solution to this problem. From my point of view it would be a lot easier if either at installation several themes based on Frameworks are already installed or if I can at installation select the first theme. These different themes could then be e.g "Bootstrap based theme" or "Tailwind CSS based theme", etc. The webmaster / developer could then take it from there and create the design he wants/needs by e.g creating a copy of the theme or by writing a "delta theme".

There are so many good CSS frameworks out there and I think contao is at the moment always behind with its own provided framework. Plus a lot of developers who would like to switch over to contao already know one of these famous CSS frameworks. So it is easier for them to start.

Another approach this is is that when you create a new theme you get asked on what it should be based and the developer can then choose e.g. "nothing", "bootstrap", "tailwind CSS", etc. And then the new theme is filled with some default values/content based on the selected CSS framework.

**Comment by aschempp** [comment]
There are extensions e.g. for Bootstrap, did you try them?

**Comment by modir** [comment]
> There are extensions e.g. for Bootstrap, did you try them?

Yes, I did. Some of them are outdated, some have other problems.

The main reason why I proposed my solution is, we have at the moment not a good and easy starting point. With WordPress it is a lot easier. You start with one of the free themes and then you adapt it to your needs. With my solution this would be solved as you could say I want to start with e.g. bootstrap or tailwind etc.

Ideally there would be an easy workflow at the beginning right after Contao is installed where it asks how you want to start with the design. As initially mentioned you can then select either one of the frameworks or "empty". If this is then solved in the background with extensions or not is for me secondary. But it would be great then if the currently shipped CSS frameworks become extensions too so that Contao is totally empty.

Specifically now about bootstrap: I could imagine that all these [Customize-functionality](https://getbootstrap.com/docs/5.3/customize/overview/) is easier to integrate if it is a theme than with an extension.

**Comment by aschempp** [comment]
> > There are extensions e.g. for Bootstrap, did you try them?
> 
> Yes, I did. Some of them are outdated, some have other problems.

So you're actual problem is that what you want does exist, but is outdated? You would prefer if it was maintained in the Contao core, because then it would always be _up to date_? I don't think that's gonna work out though.

Also, your example of Wordpress is lacking, because they ship with a default theme (as far as I know), same as Contao does with the demo.

**Comment by modir** [comment]
> So you're actual problem is that what you want does exist, but is outdated? You would prefer if it was maintained in the Contao core, because then it would always be _up to date_? I don't think that's gonna work out though.

No, my  actual problem is: Even when I remove the demo theme I still have some frontend files around. As an example:
![Screenshot from 2024-01-26 17-54-33](https://github.com/contao/contao/assets/2803913/d84fcfa2-4f3b-4bdd-9b88-a237262ec7ff)

All these frameworks and stylesheets are still laying around even though I don't need them. For me each of these CSS Framework should be one Theme each on which I can then build upon.

> Also, your example of Wordpress is lacking, because they ship with a default theme (as far as I know), same as Contao does with the demo.

Yes, but I can remove them and then all the files (CSS, JS, etc.) related to it are removed. From my example above I can not remove "12-Spalten Grid" from the system. It just stays there. (And could become a security risk in some way.)

**Comment by zoglo** [comment]
> Yes, but I can remove them and then all the files (CSS, JS, etc.) related to it are removed. From my example above I can not remove "12-Spalten Grid" from the system. It just stays there. (And could become a security risk in some way.)

Unsure how a `CSS file` could become a security risk **if you don't have it selected**. It doesn't get used, it won't even be loaded in the frontend?

See it for yourself:
https://github.com/contao/contao/blob/3600b2075906f11b930c5c10b63247c6c3a719e1/core-bundle/contao/pages/PageRegular.php#L461-L473
https://github.com/contao/contao/blob/3600b2075906f11b930c5c10b63247c6c3a719e1/core-bundle/contao/library/Contao/Controller.php#L833-L839

Many people use the CSS Frameworks when building their website and due to BC promise, you'd always have to either keep it in the core or split it into a CSS-Framework bundle. I don't see an issue with just having it there for now; there are open issues about it. I'd really love to see some proper integration of some of the most popular CSS-Frameworks, but that shouldn't be in the core.

**I don't wanna be forced to use BootStrap; I still wanna decide which one I wanna use, or if I even wanna use any amongst those.**

---

### Building websites on an existing solution

There's the official demo that you can use as a starting point if you don't remove the copyright; there are also a bunch of free themes available that you can build on.

---

### About the mentioned frameworks and comparison to WordPress (Plugins)

Imho, Tailwind is the wrong framework to start with (as it's not built to work out of the box, unless you want 2MB of CSS - Please use a CDN for that so people don't download overhead). That's for JavaScript applications as it compiles the CSS based on your needs. 
Bootstrap needs a specific structure and you are limited to what you could use, meaning that you'll most likely change the structure so it will never work with any other framework.
Afaik, Contao Bootstrap does support Bootstrap 5.3 now. WordPress wasn't built for Bootstrap nor Tailwind or anything else and relies and runs on Plugins as well to add these.

What prohibits you from overriding all templates into a bootstrap or framework-structure yourself and building your future websites on these? I don't think it's a CSS-Framework or toolset problem.

**Comment by zoglo** [comment]
> (And could become a security risk in some way.)

If it's about the pre-selected frameworks, you can always use a load_callback on the 'framework' field in 'tl_layout', modify it and return an empty array: https://docs.contao.org/dev/reference/dca/callbacks/#fields-field-load

**Comment by modir** [comment]
> Unsure how a `CSS file` could become a security risk **if you don't have it selected**. It doesn't get used, it won't even be loaded in the frontend?

The problem is that hackers could get in an other way into the database and active these files. A normal user would not know why they are active or not because they don't have enough knowledge and they trust the CMS system. And then such things could happen: https://owasp.org/www-project-web-security-testing-guide/v41/4-Web_Application_Security_Testing/11-Client_Side_Testing/05-Testing_for_CSS_Injection

> Many people use the CSS Frameworks when building their website and due to BC promise, you'd always have to either keep it in the core or split it into a CSS-Framework bundle. I don't see an issue with just having it there for now; there are open issues about it. I'd really love to see some proper integration of some of the most popular CSS-Frameworks, but that shouldn't be in the core.

That's exactly what I am saying. They should not be in the core. They should be part of the theme which you load in when needed.

> **I don't wanna be forced to use BootStrap; I still wanna decide which one I wanna use, or if I even wanna use any amongst those.**

That's exactly my point. I don't want that anybody is forced to anything. There should be many "base themes" on which you can build upon. Especially new web designers can then start with what they already know.

> Imho, Tailwind is the wrong framework to start with (as it's not built to work out of the box, unless you want 2MB of CSS - Please use a CDN for that so people don't download overhead). That's for JavaScript applications as it compiles the CSS based on your needs. Bootstrap needs a specific structure and you are limited to what you could use, meaning that you'll most likely change the structure so it will never work with any other framework. Afaik, Contao Bootstrap does support Bootstrap 5.3 now. WordPress wasn't built for Bootstrap nor Tailwind or anything else and relies and runs on Plugins as well to add these.

The question is not which framework is the best. Over time frameworks come and go. The question is what is the "easiest" system for the CMS to totally switch from one framework to another. And this is ideally done when absolutely nothing is provided by the CMS but all is in some package/theme.

> What prohibits you from overriding all templates into a bootstrap or framework-structure yourself and building your future websites on these? I don't think it's a CSS-Framework or toolset problem.

This is the wrong kind of question. In software development it is always either time or money that prohibits things. I can even extend MS Word with a huge plugin and make it a CMS. The question is how can a software (in this case the Contao CMS) make each component in the easiest way exchangeable. And this in it whole. Because what is not touched yet is the customization of the CSS framework. E.g for Bootstrap it should ask me if I want to activate additional features like e.g Bootstrap Icons. But this is dependent on the CSS Framework. And hence it would make sense that themes become "packages" where a web designer can then enable additional features like the mentioned example Bootstrap Icons.

**Comment by modir** [comment]
> > (And could become a security risk in some way.)
> 
> If it's about the pre-selected frameworks, you can always use a load_callback on the 'framework' field in 'tl_layout', modify it and return an empty array: https://docs.contao.org/dev/reference/dca/callbacks/#fields-field-load

I am not searching for the any solution. I am interested in providing a solution which:

- is more beginner friendly than now as beginners can choose what they already know and start building upon
- easier to test as such theme packages could be unit tested
- easier to automate by e.g have a build system that downloads once per week the new CSS framework version and creates a new theme version
- (And because of the point mentioned above less a security risk)
- easier to switch to a new default theme as the CMS itself doesn't need to be touched. It is only an exchange of the theme

I could mention some more points but I would say those are enough to understand my point.

**Comment by zoglo** [comment]
> This is the wrong kind of question. In software development it is always either time or money that prohibits things.

This is enough to understand what you are talking about

> That's exactly what I am saying. They should not be in the core.

This could be removed as a hard dependency in the core and splitting this apart could be contributed to Contao 5.4

> The question is not which framework is the best. Over time frameworks come and go.

Which means, they should be in plugins

> I am not searching for the any solution. I am interested in providing a solution which....

This is open source, feel free to contribute to it with pull requests regarding the issues you are mentioning. If you are not able to provide this and still want changes to happen faster: https://contao.org/de/finanziell-unterstuetzen

> The problem is that hackers could get in an other way into the database and active these files

The moment they are in your system, CSS files are not your only problem, you should know this based on your experience ;)

**Comment by modir** [comment]
> This is open source, feel free to contribute to it with pull requests regarding the issues you are mentioning. If you are not able to provide this and still want changes to happen faster: https://contao.org/de/finanziell-unterstuetzen

I would be willing to contribute the code. But as it is a) a concept change how the CMS is used and b) going quite deep into the core, I would like to have the commitment from the core team that this is the way to go. I am not going to spend many weekends for this and then it gets declined. That's why I opened this issue first to have it discussed and "approved".

> The moment they are in your system, CSS files are not your only problem, you should know this based on your experience ;)

True. But it can be something that is harder to spot once you do a clean-up of an infected system. And this I know based on my experience of cleaning up many different systems. ;)

**Comment by zoglo** [comment]
> True. But it can be something that is harder to spot once you do a clean-up of an infected system

If this is still happening, make sure to report it as a security issue (if it happened using Contao).

> a) a concept change how the CMS is used
> b) going quite deep into the core, I would like to have the commitment from the core team that this is the way to go.

Imho, not a concept change, just something to change long term; splitting this apart and adding new frameworks.

> I am not going to spend many weekends for this and then it gets declined. That's why I opened this issue first to have it discussed and "approved".

I agree with you on getting rid of the CSS frameworks as a static dependency in the code as it's hard-coded into the core right now. I don't think it being a priority since it's such a nieche issue. If you do contribute via PR, I am sure that this could change things though.

I contribute on my free time as well, I do spend many hours on my weekends to (F)OSS, sometimes nobody cares about the changes, sometimes they do, live with it or pay for the changes to be implemented since it's free open source ¯ \_(ツ)_/¯.

**Comment by aschempp** [comment]
I think we all agree the current "Contao CSS Framework" is outdated, but it's an entry-point for first users to get started quickly. We sure do want to "upgrade" it at some point, but we also need to make sure all existing sites that use this will continue to run. I don't think we'll be able to change this before Contao 6.

**Comment by modir** [comment]
> I think we all agree the current "Contao CSS Framework" is outdated, but it's an entry-point for first users to get started quickly. We sure do want to "upgrade" it at some point, but we also need to make sure all existing sites that use this will continue to run. I don't think we'll be able to change this before Contao 6.

That's why I proposed to make this outdated CSS framework a full theme which can be removed in total. Then it can still be sent along as a first theme but in can be removed as a whole.

And yes, I think as well that it is a major change that would create a new major version. That's why I wrote that it needs the commitment from the whole core team.

**Comment by zoglo** [comment]
Rewriting the hardcoded CSS-Frameworks would be the first step. 
Could move them into a component that is required (they already are -> contao-components/contao) and register them via config.yaml + dropping that dependency in Contao 6.

Registration via config could be used to drop them and register own framework files (considering they are precompiled CSS).

That could work as early as Contao 5.4 + leaving a deprecation message for Contao 6.
Also means we could add the new frameworks as well by labeling the other ones (Legacy).

Obviously this would only trigger on new installations and wouldn't need a migration.

**Comment by modir** [comment]
@zoglo I don't know the internals of Contao that well but it sounds reasonable what you wrote. This way we have a first big step done and we can then move on to something new.

**Comment by zoglo** [comment]
Since we are going full Twig Layout with slots, I don't think this is feasible anymore - At the time, Bootstrap 6.0 didn't yet drop and there is a long list of many frameworks that are dependent on build time + Code scanning, which imho is the better concept in the future.

**Comment by modir** [comment]
> Since we are going full Twig Layout with slots, I don't think this is feasible anymore - At the time, Bootstrap 6.0 didn't yet drop and there is a long list of many frameworks that are dependent on build time + Code scanning, which imho is the better concept in the future.

OK. Given the fact that I had a year ago a stroke and therefore don't have time/energy to work on this plus taking your point into consideration, I am going to close this issue.
